workspaces/src/error/impls.rs

```diff
@@ -0,0 +1,199 @@
+    where
+        T: Into<Cow<'static, str>>,
+    {
+        Self {
+            repr: ErrorRepr::Message {
+                kind,
+                message: msg.into(),
+            },
+        }
+    }
+
+    pub(crate) fn simple(kind: ErrorKind) -> Self {
+        Self {
+            repr: ErrorRepr::Simple(kind),
+        }
+    }
+
+    /// Returns the corresponding [`ErrorKind`] for this error.
+    pub fn kind(&self) -> &ErrorKind {
+        match &self.repr {
+            ErrorRepr::Simple(kind) => kind,
+            ErrorRepr::Message { kind, .. } => kind,
+            ErrorRepr::Custom { kind, .. } => kind,
+            ErrorRepr::Full { kind, .. } => kind,
+        }
+    }
+
+    /// Consumes the `Error`, returning its inner error (if any).
+    ///
+    /// If this [`Error`] was constructed via [`Error::custom`] or [`Error::full`]
+    /// then this function will return [`Some`], otherwise it will return [`None`].
+    pub fn into_inner(self) -> Result<Box<dyn std::error::Error + Send + Sync>, Self> {
+        match self.repr {
+            ErrorRepr::Custom { error, .. } => Ok(error),
+            ErrorRepr::Full { error, .. } => Ok(error),
+            _ => Err(self),
+        }
+    }
+
+    /// Attempt to downgrade the inner error to `E` if any.
+    ///
+    /// Returns `Err(self)` if the downcast is not possible
+    pub fn downcast<E: std::error::Error + Send + Sync + 'static>(self) -> Result<E, Self> {
+        if self.downcast_ref::<E>().is_none() {
+            return Err(self);
+        }
+        // Unwrapping is ok here since we already check above that the downcast will work
+        Ok(*self
+            .into_inner()?
+            .downcast()
+            .expect("failed to unwrap downcast"))
+    }
+
+    /// Returns a reference to the inner error wrapped by this error (if any).
+    pub fn get_ref(&self) -> Option<&(dyn std::error::Error + Send + Sync + 'static)> {
+        match &self.repr {
+            ErrorRepr::Custom { error, .. } => Some(error.as_ref()),
+            ErrorRepr::Full { error, .. } => Some(error.as_ref()),
+            _ => None,
+        }
+    }
+
+    /// Downcast this error object by reference.
+    pub fn downcast_ref<E: std::error::Error + Send + Sync + 'static>(&self) -> Option<&E> {
+        self.get_ref()?.downcast_ref()
+    }
+
+    /// Returns a mutable reference to the inner error wrapped by this error (if any).
+    pub fn get_mut(&mut self) -> Option<&mut (dyn std::error::Error + Send + Sync + 'static)> {
+        match &mut self.repr {
+            ErrorRepr::Custom { error, .. } => Some(error.as_mut()),
+            ErrorRepr::Full { error, .. } => Some(error.as_mut()),
+            _ => None,
+        }
+    }
+
+    /// Returns a mutable reference to the inner error (if any) downcasted to the type provided
+    pub fn downcast_mut<T: std::error::Error + Send + Sync + 'static>(&mut self) -> Option<&mut T> {
+        self.get_mut()?.downcast_mut()
+    }
```

Comment:
Do we need these in the first version? I'd probably stick to just proving `std::error::Error` and add further downcasting in follow ups, if at all necessary. 

I do think these are useful, and should be added eventually, but they also seem not really necessary for the first cut